### ISSUE TITLE
Lab 7 Discussion

### ISSUE THREAD

**Issue by tjdbryan** [nus-cs2030/2223-s2#581]
Thought everything was going well for this lab... Guess not 😰

Here's a list of constraints:
1. Cannot use `null`
2. Only one `IllegalArgumentException`
3. Only one `orElseThrow`
4. Cannot use `throw`
5. Cannot use `isEmpty()` or `isPresent()`

Anyone managed to write the `of` method with the constraints? I always end up with two `IllegalArgumentException`s or using `throw`... 😭

**Comment by choongzhanhong** [comment]
Ikr. I can't believe during the Lab I thought I got the output happy already. Went back and spent much, much, more time on Level-1 (because of this) than the other levels combined.

According to the prompt:

>  Tip: define the of(T,String) method with
> ```java
> static <T> Log<T> of(T value, String log) {
>     return Optional.ofNullable...;
> }
> ```

I'm not sure if there's a better way, but anyway, I took this literally and began my method with that line, chaining the rest with `map` and `flatMap`, finally ending off with a `orElseThrow` (There is no other way to end it because I think you MUST throw it this way).

basically did it like it's a stream and let the Optional handle the null (if any). I'm not sure if optional can take in any BiFunction (To handle both value and log together) so I just chained `map`.

**Comment by tjdbryan** [comment]
> Ikr. I can't believe during the Lab I thought I got the output happy already. Went back and spent much, much, more time on Level-1 (because of this) than the other levels combined.
> 
> According to the prompt:
> 
> > Tip: define the of(T,String) method with
> > ```java
> > static <T> Log<T> of(T value, String log) {
> >     return Optional.ofNullable...;
> > }
> > ```
> 
> I'm not sure if there's a better way, but anyway, I took this literally and began my method with that line, chaining the rest with `map` and `flatMap`, finally ending off with a `orElseThrow` (There is no other way to end it because I think you MUST throw it this way).
> 
> basically did it like it's a stream and let the Optional handle the null (if any). I'm not sure if optional can take in any BiFunction (To handle both value and log together) so I just chained `map`.

Ohh, did you do `Optional.ofNullable()` for both `value` and `log`?

**Comment by tjdbryan** [comment]
> the .of(T) can just call the other .of(T, String) using an empty ""? So only 1 of them will need to throw an exception.

Yups! I think that is the intention... But I used a if-loop for the original `of`... What was your approach to create `of`?

**Comment by timothysashimi** [comment]
> the .of(T) can just call the other .of(T, String) using an empty ""? So only 1 of them will need to throw an exception.

idts eh the "" is still considered a string and not a null value

**Comment by Hannybunny2001** [comment]
Hi, you can call the two parameter static method by calling Log.<U>of(value, "");. Ensure that you call the static method instead of the constructor

**Comment by hengbl** [comment]
> the .of(T) can just call the other .of(T, String) using an empty ""? So only 1 of them will need to throw an exception.

not sure if I'm interpreting this right, but if you meant to just return Log.<T>of(t, "") in the method .of(T t) then yep it works that's what I did.

**Comment by e0968868** [comment]
@hengbl sorry can i ask how you did this? I'm getting an error when I write this after my filter 
`.map(x -> Log<T>.of(x, ""));`

**Comment by kevincgm** [comment]
> @hengbl sorry can i ask how you did this? I'm getting an error when I write this after my filter `.map(x -> Log<T>.of(x, ""));`

seconded, I keep getting compilation error and am not sure how to deal with them..
![image](https://user-images.githubusercontent.com/93082361/227728803-c249375d-e2bd-456f-887f-7ecfec8e5ce2.png)

**Comment by hengbl** [comment]
> > @hengbl sorry can i ask how you did this? I'm getting an error when I write this after my filter `.map(x -> Log<T>.of(x, ""));`
> 
> seconded, I keep getting compilation error and am not sure how to deal with them.. ![image](https://user-images.githubusercontent.com/93082361/227728803-c249375d-e2bd-456f-887f-7ecfec8e5ce2.png)

There is no need to call map. Just return Log<T>.of(x, "") straight will do. If you think about it, Log.of(t) is the same as Log.of(t, "") right that's why you can just do that.
